src/util/taproot.rs

```diff
@@ -82,12 +92,753 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+            while start < sl.len() {
+                let elem = sha256::Hash::from_slice(&sl[start..start + TAPROOT_CONTROL_NODE_SIZE])
+                    .expect("must be multiple for 32");
+                inner.push(elem);
+                start += TAPROOT_CONTROL_NODE_SIZE;
+            }
+            Ok(TaprootMerkleBranch(inner))
+        }
+    }
+
+    /// Serialize self as bytes
+    pub fn serialize(&self) -> Vec<u8> {
+        let mut buf = Vec::with_capacity(TAPROOT_CONTROL_NODE_SIZE * self.0.len());
+        for hash in self.0.iter() {
+            buf.extend(hash.as_inner());
+        }
+        buf
+    }
+
+    // Internal function to append elements to proof
+    fn push(&mut self, h: sha256::Hash) -> Result<(), TaprootBuilderError> {
+        self.0.push(h);
+        if self.0.len() > TAPROOT_CONTROL_MAX_NODE_COUNT {
+            return Err(TaprootBuilderError::InvalidMerkleTreeDepth(self.0.len()));
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Create a MerkleProof from Vec<[`sha256::Hash`]>. Returns an error when
+    /// inner proof len is more than TAPROOT_CONTROL_MAX_NODE_COUNT (128)
+    pub fn from_inner(inner: Vec<sha256::Hash>) -> Result<Self, TaprootError> {
+        if inner.len() > TAPROOT_CONTROL_MAX_NODE_COUNT {
+            return Err(TaprootError::InvalidMerkleTreeDepth(inner.len()));
+        } else {
+            Ok(TaprootMerkleBranch(inner))
+        }
+    }
+
+    /// Consume Self to get Vec<[`sha256::Hash`]>
+    pub fn into_inner(self) -> Vec<sha256::Hash> {
+        self.0
+    }
+}
+
+/// Control Block data structure used in Tapscript satisfaction
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct ControlBlock {
+    /// The internal key
+    pub internal_key: schnorr::PublicKey,
+    /// The parity of the output key(NOT THE INTERNAL KEY WHICH IS ALWAYS XONLY)
+    pub output_key_parity: bool,
+    /// The tapleaf version,
+    pub leaf_version: LeafVersion,
+    /// The merkle proof of a script associated with this leaf
+    pub merkle_branch: TaprootMerkleBranch,
+}
+
+impl ControlBlock {
+    /// Obtain a ControlBlock from size. This is an extra witness element
+    /// that provides the proof that taproot script pubkey is correctly computed
+    /// with some specified leaf hash. This is the last element in
+    /// taproot witness when spending a output via script path.
+    ///
+    /// # Errors:
+    /// - If the control block size is not of the form 33 + 32m where
+    /// 0 <= m <= 128, InvalidControlBlock is returned
+    pub fn from_slice(sl: &[u8]) -> Result<ControlBlock, TaprootError> {
+        if sl.len() < TAPROOT_CONTROL_BASE_SIZE
+            || sl.len() > TAPROOT_CONTROL_MAX_SIZE
+            || (sl.len() - TAPROOT_CONTROL_BASE_SIZE) % TAPROOT_CONTROL_NODE_SIZE != 0
+        {
+            return Err(TaprootError::InvalidControlBlockSize(sl.len()));
+        }
+        let output_key_parity = (sl[0] & 1) == 1;
```

Comment:
Turning the slice into iterator would make it easier to reason about this code IMO, but we really need the extension trait I mentioned somewhere already.

Reply:
How do you suggest approaching this? I don't see a way to process an iterator u8 and ask it to yield slices without explicating allocating expected size arrays.